File: services/30-ws-ruby/lib/ws_auth.rb
require_relative 'jwt_service'
require_relative '../app/models/user'

# WebSocket auth — validates auth on the HTTP upgrade request.
#
# WebSocket clients cannot set custom headers in browser WebSocket API.
# Two auth paths are supported:
#
#   1. Authorization: Bearer <JWT> in the upgrade headers (server clients).
#   2. ?token=<JWT> query parameter in the WebSocket URL (browser clients).
#
# On valid auth: returns User record.
# On missing or invalid auth: returns nil. Caller closes the connection.

module WsAuth
  def self.authenticate(env)
    try_jwt(env) || try_token_param(env) || try_api_key(env)
  end

  private_class_method def self.try_jwt(env)
    header = env['HTTP_AUTHORIZATION']
    return nil unless header&.start_with?('Bearer ')

    token = header.sub('Bearer ', '')
    payload = JwtService.decode(token)
    return nil unless payload

    User.find_by(id: payload[:sub])
  end

  private_class_method def self.try_token_param(env)
    # Parse ?token= from the upgrade URL query string.
    query = env['QUERY_STRING'] || ''
    params = URI.decode_www_form(query).to_h
    token = params['token']
    return nil unless token

    payload = JwtService.decode(token)
    return nil unless payload

    User.find_by(id: payload[:sub])
  end

  private_class_method def self.try_api_key(env)
    key = env['HTTP_X_API_KEY']
    return nil unless key&.present?

    User.find_by(api_key: key)
  end
end

File: services/30-ws-elixir/lib/ws_elixir/auth/ws_auth.ex
defmodule WsElixir.Auth.WsAuth do
  @moduledoc """
  WebSocket auth — validates auth on the HTTP upgrade request.

  WebSocket browser clients cannot set custom headers.
  Two auth paths are supported:

    1. Authorization: Bearer <JWT> in the upgrade headers (server clients).
    2. ?token=<JWT> query parameter in the WebSocket URL (browser clients).

  On valid auth: returns {:ok, user_claims}.
  On missing or invalid auth: returns {:error, reason}.
  Caller closes the connection on error.

  Usage in socket_handler init/2:
    case WsAuth.authenticate(req) do
      {:ok, claims}   -> {:cowboy_websocket, req, %{user: claims}}
      {:error, _}     -> {:ok, req, :stop}
    end
  """

  alias WsElixir.Auth.Jwt

  @doc """
  Extracts and verifies auth from a Cowboy WebSocket upgrade request.

  Returns {:ok, claims} on success.
  Returns {:error, reason} on failure.
  """
  def authenticate(req) do
    with :error <- try_bearer_header(req),
         :error <- try_token_param(req) do
      {:error, "Authentication required. Provide Bearer token or ?token= parameter."}
    end
  end

  defp try_bearer_header(req) do
    case :cowboy_req.header("authorization", req) do
      <<"Bearer ", token::binary>> ->
        case Jwt.verify_token(token) do
          {:ok, claims} -> {:ok, claims}
          {:error, _}   -> :error
        end
      _ ->
        :error
    end
  end

  defp try_token_param(req) do
    qs = :cowboy_req.qs(req)
    params = URI.decode_query(qs)

    case Map.get(params, "token") do
      nil ->
        :error
      token ->
        case Jwt.verify_token(token) do
          {:ok, claims} -> {:ok, claims}
          {:error, _}   -> :error
        end
    end
  end
end
